ERCS/erc-7771.md

```diff
@@ -0,0 +1,647 @@
+
+## Backwards Compatibility
+
+An upgradeable system that does not use the router proxy, can begin to do so by simply setting a router as its single implementation, as long as the new router architecture respects the previous implementation’s storage layout.
+
+Similarly, a system that uses a router proxy can stop doing so, for example if a more advanced way to merge contracts and overcome the 24KB contract size limitation is used, such as the Smart Contract Hub proposed by maxsam4, Solidity issue #11102. If such a hub exists in the future, the system’s main proxy implementation can simply be swapped for the hub. Again, the new hub must respect the previous implementation’s storage layout.
+
+## Reference Implementation
+
+### Router Code Generation
+
+1. **Collect al system modules**
+
+   This can be done by placing all module smart contracts within a single folder, or by using some sort of nomenclature such as `*Module`.
+
+2. **Perform storage validations**
+
+   Check for storage collisions between the module’s current storage definitions.
+
+   If the system is upgradeable, check for invalid storage mutations within each module’s storage structs.
+
+3. **Collect all function selectors**
+
+   Here’s where non-duplicate selectors must be guaranteed.
+
+```
+function _validateSelectors(selectors: FunctionSelector[]) {
+  const repeated = new Set(selectors.map((s) => s.selector).filter(_onlyRepeated));
+
+  if (!repeated.size) return;
+
+  const list = selectors
+    .filter((s) => repeated.has(s.selector))
+    .map((s) => `  ${s.selector} // ${s.contractName}.${s.name}()`)
+    .join('\n');
+
+  throw new ContractValidationError(
+    `The following contracts have repeated function selectors behind the same Router:\n${list}\n`
+  );
+}
+```
+
+Selectors must then be ordered numerically,
+
+```
+selectors.sort((a, b) => {
+  return Number.parseInt(a.selector, 16) - Number.parseInt(b.selector, 16);
+});
+```
+
+4. **Construct the binary lookup table**
+
+   The binary search in the lookup table can be implemented with a simple recursive function
+
+```
+function binarySplit(node: BinaryData) {
+  if (node.selectors.length > maxSelectorsPerSwitchStatement) {
+    const midIdx = Math.ceil(node.selectors.length / 2);
+
+    const childA = binarySplit({
+      selectors: node.selectors.splice(0, midIdx),
+      children: [],
+    });
+
+    const childB = binarySplit({
+      selectors: node.selectors.splice(-midIdx),
+      children: [],
+    });
+
+    node.children.push(childA);
+    node.children.push(childB);
+
+    node.selectors = [];
+  }
+
+  return node;
+}
+```
+
+5. **Render the router code**
+
+   This can be as simple as using a Mustache template to insert the module and selector data.
+
+```
+  return Mustache.render(template, {
+    moduleName: routerName,
+    modules: _renderModules(contracts),
+    selectors: _renderSelectors(binaryData),
+  });
+```
+
+With a template such as,
+
+```
+//SPDX-License-Identifier: Unlicense
```

Comment:
Code within the proposal itself must be CC0-1.0 licensed. We do permit other licenses in the assets directory.